fix(core): remove application from the testability registry when the root view is removed

arturovt opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #22106


## What is the new behavior?

In the new behavior Angular removes applications from the testability registry when the root view gets destroyed. This eliminates a memory leak, because before that the `TestabilityRegistry` holds references to HTML elements, thus they cannot be GCed.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

arturovt committed:
fix(core): remove application from the testability registry when the root view is removed

In the new behavior Angular removes applications from the testability registry when the
root view gets destroyed. This eliminates a memory leak, because before that the
TestabilityRegistry holds references to HTML elements, thus they cannot be GCed.

PR Close #22106

mhevery commented:
[presubmit](http://test/OCL:344846692:BASE:344846718:1606762758423:c23ef41d)

AndrewKushnir reviewed (commented):
@arturovt, thanks for the PR, I've just left a couple comments. Thank you.

arturovt commented on packages/core/src/application_ref.ts:
```diff
@@ -721,8 +723,13 @@ export class ApplicationRef {
+      const testabilityRegistry = compRef.injector.get(TestabilityRegistry);
+      testabilityRegistry.registerApplication(compRef.location.nativeElement, testability);
+      // Actually, we could use `compRef.onDestroy` for this kind of task,
+      // but this will not work in Ivy.
```
I've taken it from that comment: https://github.com/angular/angular/issues/39365#issuecomment-714827915
Do you think it's worth putting this link into comments? I'll do if it is.

AndrewKushnir replied:
@arturovt, thanks for additional information. The `compRef.onDestroy` is used a few lines above, can we check if it gets executed or not with Ivy? If yes, we can put the `testabilityRegistry.unregisterApplication` there as well, but if it doesn't work with Ivy, we should fix that too (by creating a "local" array of onDestroy callbacks or `compRef` references), thus it might contribute to the memory leak problem.

arturovt replied:
I've checked, it's not executed in Ivy. Do you think it's worth moving `this._unloadComponent(compRef)` from `compRef.onDestroy` to `_unregisterApplicationFns`?

arturovt replied:
I've checked, it's not executed in Ivy. Updated the code.

arturovt commented on packages/core/src/application_ref.ts:
```diff
@@ -721,8 +723,13 @@ export class ApplicationRef {
-      compRef.injector.get(TestabilityRegistry)
-          .registerApplication(compRef.location.nativeElement, testability);
+      const testabilityRegistry = compRef.injector.get(TestabilityRegistry);
+      testabilityRegistry.registerApplication(compRef.location.nativeElement, testability);
```
Makes sense, updated.

mhevery commented on packages/core/src/application_ref.ts:
```diff
@@ -805,6 +812,10 @@ export class ApplicationRef {
   ngOnDestroy() {
     // TODO(alxhub): Dispose of the NgZone.
```
I don't see anything abvious. @alxhub ?

AndrewKushnir replied:
I don't see anything obviously missing too, so I'll remove this TODO.

mhevery commented:
I fixed up the code to make it consistent between Ivy and VE. @AndrewKushnir please have a look.

mhevery commented:
[presubmit](http://test/OCL:344846692:BASE:344917162:1606784379067:44b6627b)

AndrewKushnir commented:
@mhevery thanks, I will have a look and add extra tests as we discussed. I just noticed that the commit message is a bit strange: it refers to a different commit from master branch - could you please have a look when you get a chance?

@arturovt just wanted to give a quick update: we've came across some inconsistencies in existing code while reviewing the changes in this PR and we went ahead with additional fixes:

- to make sure that the cleanup is consistent in case `ApplicationRef` is destroyed (as a result of an `Injector` destroy) or a `ComponentRef.destroy` is called.
- the callbacks registered with `componentRef.onDestroy(cb)` are invoked when corresponding view is removed. This was the problem in #39365 that you've referred to. So these changes should address #39365 as well.

I plan to add a couple more tests to verify the above-mentioned scenarios (and the one from #39365) and will push an extra commit. Please let us know if you're ok with us contributing to your PR (if you're not ok, we can revert our changes and create a separate PR).

@arturovt it'd be helpful if you could also have a look at the changes in this PR (as original author of the changes) and let us know if you have any questions/feedback.

Thank you.

arturovt commented:
@AndrewKushnir I don't have any questions, looks great. Seems also great that it closes another issue simultaneously.

AndrewKushnir committed:
test(core): verify `onDestroy` callbacks are invoked when ComponentRef is destroyed

This commit adds a few tests to verify that the `onDestroy` callbacks are invoked when `ComponentRef` instance
is destroyed and the logic is consistent between ViewEngine and Ivy.

AndrewKushnir commented:
[Presubmit](http://test/OCL:345112839:BASE:345112864:1606865366971:9e944e20).

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

mhevery reviewed (approved):
reviewed-for: global-approvers
